src/compile/model.ts

```diff
@@ -268,7 +270,14 @@ export abstract class Model {
   public abstract assembleSelectionSignals(): any[];
 
   public abstract assembleSelectionData(data: VgData[]): VgData[];
-  public abstract assembleData(): VgData[];
+  public assembleData(): VgData[] {
+     if (!this.parent) {
+      // only assemble data in the root
+      optimizeDataflow(this.component.data);
```

Comment:
Hmm if we call it here, it may still happens after other things in compile.ts? 
Why don't we call it at the end of parse instead?